common/progress.go

```diff
@@ -0,0 +1,61 @@
+package common
+
+import (
+	"fmt"
+	"github.com/cheggaaa/pb"
+	"github.com/hashicorp/packer/packer"
+	"log"
+	"time"
+)
+
+// Default progress bar appearance
+func GetDefaultProgressBar(ui *packer.Ui) pb.ProgressBar {
+	bar := pb.New64(0)
+	bar.ShowPercent = true
+	bar.ShowCounters = true
+	bar.ShowSpeed = false
+	bar.ShowBar = true
+	bar.ShowTimeLeft = false
+	bar.ShowFinalTime = false
+	bar.SetUnits(pb.U_BYTES)
+	bar.Format("[=>-]")
+	bar.SetRefreshRate(1 * time.Second)
+	bar.SetWidth(80)
+
+	// If there's no UI set, then the progress bar doesn't need anything else
+	if ui == nil {
+		return *bar
+	}
+	UI := *ui
+
+	// Now check the UI's width to adjust the progress bar
+	uiWidth := UI.GetMinimumLength() + len("\n")
+
+	// If the UI's width is signed, then this interface doesn't really
+	// benefit from a progress bar
+	if uiWidth < 0 {
+		log.Println("Refusing to render progress-bar for unsupported UI.")
+		return *bar
+	}
+	bar.Callback = UI.Message
+
+	// Figure out the terminal width if possible
+	width, _, err := GetTerminalDimensions()
```

Comment:
oh, that's crazy. didn't know that. that's awesome. it looks like he did it the same way that I did too. i'll fix it in the next commit.

Reply:
👍

Reply:
Ok, correction. It turns out that `pb.GetTerminalWidth()` doesn't work on windows because the author is calling `kernel32.GetConsoleScreenBufferInfo` using `syscall.Stdout` (which I think is invalid due to packer's stdout being related to the rpc interface and not the console).

Since this handle isn't valid at the time, on windows the API call fails with an error "The handle is invalid". In mine, I open up a handle to "CONOUT$" (and then close it) which results in it always succeeding.

Reply:
He's using `ws, err := unix.IoctlGetWinsize(fd, unix.TIOCGWINSZ)` though, so I'll switch to using that instead of the direct call to `Ioctl` so that it works on darwin too.